### PR TITLE
re-introduce: stage3: efficient compute of scaled_global_grad_norm

### PR DESCRIPTION
**Pull request by nelyahu**
reverting previous revert of this feature:
https://github.com/nelyahu/DeepSpeed/commit/bc48371c5e1fb8fd70fc79285e66201dbb65679b
in addition,
bug fix for offload mode.

### CODE REVIEW

**Comment by nelyahu** [comment]
Hi @lekurile ,
can you please run ds-chat coverage on this PR? i reproduced the issue was reproted in https://github.com/nelyahu/DeepSpeed/commit/bc48371c5e1fb8fd70fc79285e66201dbb65679b  and fixed it.
would like to get a pre-commit validation on this test suit.
CC: @tjruwase

**Comment by lekurile** [comment]
> Hi @lekurile , can you please run ds-chat coverage on this PR? i reproduced the issue was reproted in [nelyahu@bc48371](https://github.com/nelyahu/DeepSpeed/commit/bc48371c5e1fb8fd70fc79285e66201dbb65679b) and fixed it. would like to get a pre-commit validation on this test suit. CC: @tjruwase

Hi @nelyahu, thank you for the PR. I've kicked off a test run here:
https://github.com/microsoft/DeepSpeed/actions/runs/8927396023

**Comment by nelyahu** [comment]
> > Hi @lekurile , can you please run ds-chat coverage on this PR? i reproduced the issue was reproted in [nelyahu@bc48371](https://github.com/nelyahu/DeepSpeed/commit/bc48371c5e1fb8fd70fc79285e66201dbb65679b) and fixed it. would like to get a pre-commit validation on this test suit. CC: @tjruwase
> 
> Hi @nelyahu, thank you for the PR. I've kicked off a test run here: https://github.com/microsoft/DeepSpeed/actions/runs/8927396023

Thanks @lekurile , seems like it passed, can you confirm?

**Comment by lekurile** [comment]
> > > Hi @lekurile , can you please run ds-chat coverage on this PR? i reproduced the issue was reproted in [nelyahu@bc48371](https://github.com/nelyahu/DeepSpeed/commit/bc48371c5e1fb8fd70fc79285e66201dbb65679b) and fixed it. would like to get a pre-commit validation on this test suit. CC: @tjruwase
> > 
> > 
> > Hi @nelyahu, thank you for the PR. I've kicked off a test run here: https://github.com/microsoft/DeepSpeed/actions/runs/8927396023
> 
> Thanks @lekurile , seems like it passed, can you confirm?

Yep, looks like it passed, approved the PR and running all checks.

**Review by lekurile — APPROVED** [review]

File: deepspeed/runtime/zero/stage3.py [hunk 1]
```diff
@@ -1409,7 +1409,7 @@ def complete_grad_norm_calculation_for_cpu_offload(self, params):
         norm_is_nan = total_norm.isnan()
         inf_or_nan = norm_is_nan.logical_or(norm_is_inf)
 
-        err = torch.tensor(-1.0, device=self.device, dtype=torch.float)
+        err = torch.tensor(-1.0, device=inf_or_nan.device, dtype=torch.float)
```
[the remark this answers is not in the document]
Comment (nelyahu) [reply deepspeed/runtime/zero/stage3.py:1412]: @lekurile Yes, this PR includes 2 commits (original PR, and the fix). this specific line change is part of the fix for the bug.

**Pull request merged**